ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp

```diff
@@ -8707,3 +8709,150 @@ int OGRGeoPackageTableLayer::GetNextArrowArrayInternal(
 
     return 0;
 }
+
+/************************************************************************/
+/*               OGR_GPKG_GeometryExtent3DAggregate()                   */
+/************************************************************************/
+
+namespace
+{
+struct GeometryExtent3DAggregateContext
+{
+    sqlite3 *m_hDB = nullptr;
+    int m_nFlags = 0;
+    OGREnvelope3D m_oExtent3D;
+
+    explicit GeometryExtent3DAggregateContext(sqlite3 *hDB, int nFlags)
+        : m_hDB(hDB), m_nFlags(nFlags)
+    {
+    }
+    GeometryExtent3DAggregateContext(const GeometryExtent3DAggregateContext &) =
+        delete;
+    GeometryExtent3DAggregateContext &
+    operator=(const GeometryExtent3DAggregateContext &) = delete;
+};
+
+}  // namespace
+
+static void OGR_GPKG_GeometryExtent3DAggregate_Step(sqlite3_context *pContext,
+                                                    int /*argc*/,
+                                                    sqlite3_value **argv)
+{
+    const GByte *pabyBLOB =
+        reinterpret_cast<const GByte *>(sqlite3_value_blob(argv[0]));
+
+    auto poContext = static_cast<GeometryExtent3DAggregateContext *>(
+        sqlite3_user_data(pContext));
+
+    if (pabyBLOB != nullptr)
+    {
+        GPkgHeader sHeader;
+        const int nBLOBLen = sqlite3_value_bytes(argv[0]);
+        if (GPkgHeaderFromWKB(pabyBLOB, nBLOBLen, &sHeader) == OGRERR_NONE &&
+            static_cast<size_t>(nBLOBLen) >= sHeader.nHeaderLen + 5)
+        {
+            if (!OGRGeoPackageGetHeader(pContext, 0, argv, &sHeader, true))
+                return;
+        }
+        OGREnvelope3D extent3D;
+        extent3D.MinX = sHeader.MinX;
+        extent3D.MaxX = sHeader.MaxX;
+        extent3D.MinY = sHeader.MinY;
+        extent3D.MaxY = sHeader.MaxY;
+        extent3D.MinZ = sHeader.MinZ;
+        extent3D.MaxZ = sHeader.MaxZ;
```

Comment:
There's one case that is probably not well covered. It is possible that a 3D geometry has a extent in the GPKG geometry header, but without the Z values (. sHeader.MinZ/sHeader.MaxZ as returned by GPkgHeaderFromWKB() are only valid if poHeader->bExtentHasZ .
Cf http://www.geopackage.org/spec/#gpb_format and bit E: envelope contents indicator code (3-bit unsigned integer).
I don't think the GeoPackage spec mandates to write a 3D envelope if the geometry is 3D (the GDAL GPKG writer will however always write a 3D envelope for a 3D geom, so this would be for .gpkg created by othersoftware). It could be just 2D.
So I believe OGRGeoPackageGetHeader() should be modified to have an extra bNeedExtent3D flag, where you would call getEnvelope3D() if only having a 2D envelope in the GeoPackage geometry header.

Reply:
I see... Do you have any recommendation about how to build a test for that case? 

Also it is not entirely clear to me when this branch is triggered (and how to test it): https://github.com/OSGeo/gdal/pull/8910/files#diff-1b62be67e99a70d80f3b1651237e60b85c47cdb406a591e5d50dd0f13676361bR529

Reply:
> Do you have any recommendation about how to build a test for that case?

the easiest would probably to patch GPkgGeometryFromOGR(), around lines 292-347 for the purposes of generating test datasets. One without any envelope at all, and another one with 2D envelopes

Reply:
> Also it is not entirely clear to me when this branch is triggered (and how to test it): https://github.com/OSGeo/gdal/pull/8910/files#diff-1b62be67e99a70d80f3b1651237e60b85c47cdb406a591e5d50dd0f13676361bR529

perhaps issue an ``UPDATE table_name SET geom = 'some garbage here'`` on an existing .gpkg ?

Reply:
ok, done.